### PR TITLE
First implementation opfab API

### PR DESCRIPTION
**Pull request by freddidierRTE**
- In release note :
  -  In chapter :  Features 
  -  Text : #4438 Opfab API V1

### CODE REVIEW

**Review by quinarygio — CHANGES_REQUESTED** [review]

File: src/docs/asciidoc/reference_doc/ui_api.adoc [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+It is possible to be registered when opfab has finished all tasks regarding rendering template by setting a listener function .The function will be called after the call of the other listener (applyChildCard, lockAnswer ,lttdExpired and screenSize)
+
+It can be used by a template to launch some processing when loading is complete
+
+```
+        opfab.currentCard.listenToTemplateRenderingComplete(() => {// do some stuff})
+```
+
+=== Function registerGetUserResponse
```
Comment (vlo-rte) [inline src/docs/asciidoc/reference_doc/ui_api.adoc:240]: registerFunctionToGetUserResponse

File: src/docs/asciidoc/resources/migration_guide_to_4.0.adoc [hunk 2]
```diff
@@ -20,4 +20,77 @@ Now `ui-menu.json` file contains 4 fields :
...
+|opfab.currentCard.displayLoadingSpinner()
+
+|templateGateway.hideLoadingSpinner()
+|opfab.currentCard.hideLoadingSpinner()
+
+|templateGateway.childCards
+|opfab.currentCard.getChildCards()
+
+|templateGateway.isLocked
+|opfab.currentCard.isLocked()
```
Comment (vlo-rte) [inline src/docs/asciidoc/resources/migration_guide_to_4.0.adoc:70]: isResponseLocked

File: src/docs/asciidoc/resources/migration_guide_to_4.0.adoc [hunk 3]
```diff
@@ -20,4 +20,77 @@ Now `ui-menu.json` file contains 4 fields :
...
+
+|templateGateway.lockAnswer = function () {//do some stuff}
+|opfab.currentCard.listenToAnswerLock( () => {//do some stuff} )
+
+|templateGateway.unlockAnswer = function () {//do some stuff}
+|opfab.currentCard.listenToAnswerUnlock( ()  => {//do some stuff} )
```
Comment (vlo-rte) [inline src/docs/asciidoc/resources/migration_guide_to_4.0.adoc:76]: listenToResponseUnlock

File: src/docs/asciidoc/resources/migration_guide_to_4.0.adoc [hunk 4]
```diff
@@ -20,4 +20,77 @@ Now `ui-menu.json` file contains 4 fields :
...
+|opfab.currentCard.listenToStyleChange( ()  => {//do some stuff} )
+
+|templateGateway.setScreenSize =  function () {//do some stuff}
+|opfab.currentCard.listenToScreenSize( ()  => {//do some stuff} )
+
+|templateGateway.onTemplateRenderingComplete =  function () {//do some stuff}
+|opfab.currentCard.listenToTemplateRenderingComplete( ()  => {//do some stuff} )
+
+|templateGateway.getUserResponse =  function () {//do some stuff}
+|opfab.currentCard.registerGetUserResponse( ()  => {//do some stuff} )
```
Comment (vlo-rte) [inline src/docs/asciidoc/resources/migration_guide_to_4.0.adoc:91]: registerFunctionToGetUserResponse

File: src/test/api/karate/businessconfig/resources/bundle_test_action/template/template1.handlebars [hunk 5]
```diff
@@ -36,7 +44,7 @@
     }
   }
 
-  templateGateway.getUserResponse = function() {
+  opfab.currentCard.registerGetUserResponse(() => {
```
Comment (vlo-rte) [inline src/test/api/karate/businessconfig/resources/bundle_test_action/template/template1.handlebars:47]: registerFunctionToGetUserResponse

File: src/test/resources/bundles/messageOrQuestionExample/template/confirmation.handlebars [hunk 6]
```diff
@@ -80,8 +80,9 @@
                 } else {
                     childsDiv.innerHTML = '';
                 }
-            }
-            templateGateway.getUserResponse = function () {
+            });
+
+            opfab.currentCard.registerGetUserResponse(() => {
```
Comment (vlo-rte) [inline src/test/resources/bundles/messageOrQuestionExample/template/confirmation.handlebars:85]: registerFunctionToGetUserResponse

File: ui/main/src/app/business/services/opfabAPI.service.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        };
+        opfab.currentCard.getEntityUsedForUserResponse = function () {
+            return self.currentCard.entityUsedForUserResponse;
+        };
+
+        opfab.currentCard.hideLoadingSpinner = function () {
+            self.currentCard.hideLoadingSpinner();
+        };
+
+        opfab.currentCard.isLocked = function () {
```
Comment (vlo-rte) [inline ui/main/src/app/business/services/opfabAPI.service.ts:157]: isResponseLocked

File: ui/main/src/assets/js/templateGateway.js [hunk 8]
```diff
@@ -8,170 +8,183 @@
...
+            new Date().toISOString(),
+            ' WARNING : Use of templateGateway.childCards is deprecated , you shall use opfab.currentCard.getChildCards() instead'
+        );
+        return opfab.currentCard.getChildCards();
+    },
+
+    get isLocked() {
+        console.warn(
+            new Date().toISOString(),
+            ' WARNING : Use of templateGateway.isLocked is deprecated , you shall use opfab.currentCard.isLocked() instead'
```
Comment (vlo-rte) [inline ui/main/src/assets/js/templateGateway.js:122]: isResponseLocked

File: ui/main/src/assets/js/templateGateway.js [hunk 9]
```diff
@@ -8,170 +8,183 @@
...
+        );
+        return opfab.currentCard.isLocked();
+    },
+
+    set lockAnswer(value) {
+        console.warn(
+            new Date().toISOString(),
+            ' WARNING : Use of templateGateway.lockAnswer is deprecated , you shall use opfab.currentCard.listenToAnswerLock() instead'
+        );
+        opfab.currentCard.listenToAnswerLock(value);
```
Comment (vlo-rte) [inline ui/main/src/assets/js/templateGateway.js:132]: listenToResponseLock

File: ui/main/src/assets/js/templateGateway.js [hunk 10]
```diff
@@ -8,170 +8,183 @@
...
+    },
+
+    set unlockAnswer(value) {
+        console.warn(
+            new Date().toISOString(),
+            ' WARNING : Use of templateGateway.unlockAnswer is deprecated , you shall use opfab.currentCard.listenToAnswerUnlock() instead'
+        );
+        opfab.currentCard.listenToAnswerUnlock(value);
```
Comment (vlo-rte) [inline ui/main/src/assets/js/templateGateway.js:140]: listenToResponseUnlock

File: ui/main/src/assets/js/templateGateway.js [hunk 11]
```diff
@@ -8,170 +8,183 @@
...
+        );
+        opfab.currentCard.listenToTemplateRenderingComplete(value);
+    },
+
+    set getUserResponse(value) {
+        console.warn(
+            new Date().toISOString(),
+            ' WARNING : Use of templateGateway.getUserResponse is deprecated , you shall use opfab.currentCard.registerGetUserResponse() instead'
+        );
+        opfab.currentCard.registerGetUserResponse(value);
```
Comment (vlo-rte) [inline ui/main/src/assets/js/templateGateway.js:180]: registerFunctionToGetUserResponse

**Review by vlo-rte — CHANGES_REQUESTED** [review]

File: src/docs/asciidoc/reference_doc/ui_api.adoc [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+// See AUTHORS.txt
+// This document is subject to the terms of the Creative Commons Attribution 4.0 International license.
+// If a copy of the license was not distributed with this
+// file, You can obtain one at https://creativecommons.org/licenses/by/4.0/.
+// SPDX-License-Identifier: CC-BY-4.0
+
+
+= Frontend API
+
+The frontend API is aimed to provide a way for template to communicate with the core opfab code. It's divided as follow :
```
Comment (ClementBouvierN) [inline src/docs/asciidoc/reference_doc/ui_api.adoc:11]: "is aimed to " -> "is intended to" / "is aiming to"

File: src/docs/asciidoc/reference_doc/ui_api.adoc [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+== Users API 
+
+=== Entities
+
+Entity object contains the following fields :
+
+- 'id' : id of the entity
+- 'name' : name of the entity
+- 'description' : description of the entity
+- 'entityAllowedToSendCard' : boolean indicating whether the entity is allowed to send card or not
```
Comment (ClementBouvierN) [inline src/docs/asciidoc/reference_doc/ui_api.adoc:26]: " to send card" -> "to send a card" / "to send cards"

File: src/docs/asciidoc/reference_doc/ui_api.adoc [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+This can be done by calling the following function from the template :
+
+```
+opfab.navigate.redirectToBusinessMenu(idMenu,idEntry);
+```
+
+- idMenu is the id of the menu defined in ui-menu.json
+- idEntry is the id of the entry defined in ui-menu.json
+
+It is also possible to append url extension (sub paths and/or parameters) to the url that will be called:
```
Comment (ClementBouvierN) [inline src/docs/asciidoc/reference_doc/ui_api.adoc:73]: " append url extension" -> "append a url extension"

File: src/docs/asciidoc/reference_doc/ui_api.adoc [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+=== Function getChildCards
+
+ Returns an array of the child cards. The structure of a child card is the same as the structure of a classic card.
+```
+const childCards =  opfab.currentCard.getChildCards(); 
+```
+
+=== Function getDisplayContext
+
+To adapt the template content to the display context it is possible to get from OperatorFabric the page context where the template will be rendered by calling the javascript function getDisplayContext(). The function returns a string with one of the following values :
```
Comment (ClementBouvierN) [inline src/docs/asciidoc/reference_doc/ui_api.adoc:117]: the page context -> the page's context

File: src/docs/asciidoc/reference_doc/ui_api.adoc [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+
+If the card has a last time to decide (lttd) configured, when the time is expired this information can be received by the template by registering a listener.
+
+```
+    opfab.currentCard.listenToLttdExpired( () => { // do some stuff });
+```
+
+=== Function listenToStyleChange
+
+Card template can be informed when switching day/night mode by registering a listener  as follow :
```
Comment (ClementBouvierN) [inline src/docs/asciidoc/reference_doc/ui_api.adoc:212]: " a listener  as " -> 2 spaces between listener and as

File: src/docs/asciidoc/reference_doc/ui_api.adoc [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+
+It can be used by a template to launch some processing when loading is complete
+
+```
+        opfab.currentCard.listenToTemplateRenderingComplete(() => {// do some stuff})
+```
+
+=== Function registerFunctionToGetUserResponse
+```
+        opfab.currentCard.registerFunctionToGetUserResponse
```
Comment (ClementBouvierN) [inline src/docs/asciidoc/reference_doc/ui_api.adoc:248]: no explanations ?

**Review by vlo-rte — APPROVED** [review]

**Pull request merged**